### PR TITLE
Add custom code at deployment of SC.

### PR DESCRIPTION
**Pull request by AurelienFT**
Fix: https://github.com/massalabs/massa-sc-toolkit/issues/52

### CODE REVIEW

File: commands/init/deployer/deployer.as.ts
Comment (gregLibert) [inline commands/init/deployer/deployer.as.ts:15]: hum, I'm not a big fan of that...

File: commands/init/deployer/compile-deployer.ts
Comment (gregLibert) [inline commands/init/deployer/compile-deployer.ts:14]: When are trying to avoid token/regexp replacer. We have seen in the past that it's can be buggy. Can you use a replacer instead?

**Review by gregLibert — CHANGES_REQUESTED** [review]
I'm not a big fan of this solution. But maybe I'm wrong.

I thought I would solve this problem in a different way by changing the way the deployer works:
- define the deployer as a smart contract on the blockchain
- call this smart contract with two arguments:
  - SC byte code
  - initialization function byte code

Wdyt ?

**Comment by gregLibert** [comment]
cc @Thykof @longgamma-massa

**Comment by AurelienFT** [comment]
> I'm not a big fan of this solution. But maybe I'm wrong.
> 
> I thought I would solve this problem in a different way by changing the way the deployer works:
> 
> * define the deployer as a smart contract on the blockchain
> * call this smart contract with two arguments:
>   
>   * SC byte code
>   * initialization function byte code
> 
> Wdyt ?

It should be possible and even better with #67 . As we could have only one SC that deploy everything pass in the operation datastore. Multiple problems I see with this approach : 

- Cost more gas because you will have to check if the argument `init_function` is setup
- Trust an SC on chain

It's not big problems and i'm up to go with your solution just I think we should do it in the same time as #67 and so, have a design that allow multiple SC from datastore to be deployed.

**Comment by gregLibert** [comment]
> * Cost more gas because you will have to check if the argument `init_function` is setup

We are just talking about a if here, aren't we ?

> * Trust an SC on chain

Good point, this is true for all SC and should be solved globally (at the block explorer level ?)

**Comment by AurelienFT** [comment]
> We are just talking about a if here, aren't we ?

Yes


> Good point, this is true for all SC and should be solved globally (at the block explorer level ?)

Yes in a way but you can deploy SC without trusting anyone currently

**Comment by AurelienFT** [comment]
Two things need to be done to be sure it works : 

- Generate version of sc deployer and add it to the CI
- Use node API addresses in the .env

File: commands/init/deployer/deployer.as.ts
Comment (AurelienFT) [reply commands/init/deployer/deployer.as.ts]: I resolve because it's outdated.

File: commands/init/deployer/compile-deployer.ts
Comment (AurelienFT) [reply commands/init/deployer/compile-deployer.ts]: I resolve because it's outdated. We use a new way now.

File: commands/init/assembly/contracts/main.ts
Comment (Thykof) [inline commands/init/assembly/contracts/main.ts:13]:
```suggestion
  const argsDeserialized = new Args(args);
  const name = argsDeserialized.nextString().unwrap();
```

File: sc-deployer/assembly/contracts/deployer.ts
Comment (Thykof) [inline sc-deployer/assembly/contracts/deployer.ts:6]: please use camel case

File: sc-deployer/index.ts
Comment (Thykof) [inline sc-deployer/index.ts:20]: what do you mean by 'this ip is false'?

File: sc-deployer/index.ts
Comment (Thykof) [inline sc-deployer/index.ts:31]: it's typescript so I guess you can go with Array.forEarch

```typescript
    contracts.forEach((contract, i) => {
```

File: sc-deployer/index.ts
Comment (Thykof) [inline sc-deployer/index.ts:38]: this is hard to read and this will be hard to maintain the Map, we should at least document its datastructure

File: sc-deployer/package.json
Comment (Thykof) [inline sc-deployer/package.json:9]: we should add a command to build the wasm files (deployer, and test) and remove from the repository

File: sc-deployer/package.json
Comment (Thykof) [inline sc-deployer/package.json:19]: we use 0.25
```suggestion
    "assemblyscript": "^0.25.1",
```

File: sc-deployer/tests/index.ts
Comment (Thykof) [inline sc-deployer/tests/index.ts:7]: warning private key here

does this test file meant to deploy something into the blockchain? it should target a sandbox node

**Review by Thykof — CHANGES_REQUESTED** [review]
Why do you commit binary wasm file and dist folder? we could build them in the CI and publish the built in the CD.

We used npm workspaces for monorepo management, you can get inspired by https://github.com/massalabs/as/pull/58/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519

File: commands/init/package.json
Comment (AurelienFT) [reply commands/init/package.json:24]: I will create the other repository and edit here before merging this one.

File: sc-deployer/package.json
Comment (AurelienFT) [reply sc-deployer/package.json:9]: The command is the build above. Removed from repository.

File: sc-deployer/index.ts
Comment (AurelienFT) [reply sc-deployer/index.ts:20]: I use the same IP as the public one so with 33035 as port so it will not work for a private API but we don't care in our case.

File: sc-deployer/assembly/contracts/deployer.ts
Comment (AurelienFT) [reply sc-deployer/assembly/contracts/deployer.ts:6]: Done.

File: commands/init/assembly/contracts/main.ts
Comment (AurelienFT) [reply commands/init/assembly/contracts/main.ts]: Done.

File: commands/init/package.json
Comment (Thykof) [reply commands/init/package.json:24]: new github repositoy? I though we agreed on make this one a monorepo, do you mean new npm package?

File: sc-deployer/index.ts
Comment (AurelienFT) [reply sc-deployer/index.ts:54]: I have added a detailed doc now. Is it better to understand ?

Comment (AurelienFT) [reply sc-deployer/index.ts:54]: I have added a detailed doc now. Is it better to understand ?

File: sc-deployer/index.ts
Comment (AurelienFT) [reply sc-deployer/index.ts]: Done.

File: sc-deployer/package.json
Comment (AurelienFT) [reply sc-deployer/package.json:19]: Done.

File: commands/init/package.json
Comment (AurelienFT) [reply commands/init/package.json:24]: I don't know what's the best to be honest :/

File: commands/init/deploy.ts
Comment (AurelienFT) [reply commands/init/deploy.ts:27]: I'm not a fan of removing the account parameter it's really useful I don't see any default value that are acceptable.
I'm ok for the last two parameters optional.
I'm also not a fan of removing `readFileSync` the people using it are developers are they have developed the contracts I think it's preferable to keep a bit of code and so we have a more flexibility and possibilities. We are making a TS file let's use the possibilities of TS otherwise we can just make a JSON that we parse and polymorph in all ways.

File: sc-deployer/tests/index.ts
Comment (AurelienFT) [reply sc-deployer/tests/index.ts:7]: Done.

File: commands/init/deploy.ts
Comment (Thykof) [reply commands/init/deploy.ts:27]: I get you points, I was thinking of simplifying the life of the developer, for example with the ability to deploy in 2 lines like in https://hardhat.org/hardhat-runner/docs/guides/deploying or https://trufflesuite.com/docs/truffle/how-to/contracts/run-migrations/

Comment (AurelienFT) [reply commands/init/deploy.ts:27]: I personally find it more complicated to understand in the first one because it's too shady. The second one have the require above. I will open an issue on that but I think for a first version we can keep it like that and decide after if that's ok for you.

File: commands/init/package.json
Comment (gregLibert) [reply commands/init/package.json:24]: The best approach here is to be consistent. Let's try the monorepo approach.
If it turns out that the adventure becomes a nightmare, there is always time to change your mind.

**Review by Thykof — COMMENTED** [review]
There is still sc-deployer/dist/deployer.wasm that you can delete

File: commands/init/package.json
Comment (Thykof) [reply commands/init/package.json:24]: good, so @AurelienFT you can setup monorepo with npm workspaces

File: commands/init/deploy.ts
Comment (Thykof) [reply commands/init/deploy.ts:27]: okay

File: commands/init/package.json
Comment (Thykof) [reply commands/init/package.json:24]: this will solve https://github.com/massalabs/massa-sc-toolkit/issues/36

File: sc-deployer/assembly/contracts/deployer.ts
Comment (AurelienFT) [reply sc-deployer/assembly/contracts/deployer.ts]: Oh yes I'm retarded. Done.

**Comment by Thykof** [comment]
In this PR I made this repo a monorepo for multiple npm packages:
- `@massalabs/sc-toolkit`,
- `@massalabs/massa-sc-deployer`,
- `@massalabs/massa-sc-compiler`

**Review by Thykof — APPROVED** [review]

**Review by AurelienFT — COMMENTED** [review]
Nice refactor @Thykof ! Thanks a lot

File: packages/sc-toolkit/commands/init/assembly/contracts/main.ts
Comment (gregLibert) [inline packages/sc-toolkit/commands/init/assembly/contracts/main.ts:11]: The best way to ensure that a function is not called in future is to not write that function :-)

File: packages/sc-toolkit/commands/init/assembly/contracts/main.ts
Comment (gregLibert) [inline packages/sc-toolkit/commands/init/assembly/contracts/main.ts:25]: Why do we need that ?

File: packages/sc-toolkit/commands/init/assembly/contracts/main.ts
Comment (gregLibert) [inline packages/sc-toolkit/commands/init/assembly/contracts/main.ts:13]: I think we should use a simpler and more meaningful alias

File: packages/sc-toolkit/commands/init/src/deploy.ts
Comment (gregLibert) [inline packages/sc-toolkit/commands/init/src/deploy.ts:18]: why a __ prefix ?

File: packages/sc-toolkit/commands/init/src/deploy.ts
Comment (gregLibert) [inline packages/sc-toolkit/commands/init/src/deploy.ts:33]: Magic numbers

File: packages/sc-toolkit/src/initialize.js
Comment (gregLibert) [inline packages/sc-toolkit/src/initialize.js:17]:
```suggestion
      `The project directory ${directory} already exists. Please choose another project name.`,
```

File: packages/sc-deployer/src/index.ts
Comment (gregLibert) [inline packages/sc-deployer/src/index.ts:60]: Coole explanation. Is it present in other repo such as mass-as-sdk ?

File: packages/sc-deployer/assembly/contracts/deployer.ts
Comment (gregLibert) [inline packages/sc-deployer/assembly/contracts/deployer.ts:11]: The readibility is to be improve...

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: packages/sc-toolkit/commands/init/README.md
Comment (Thykof) [inline packages/sc-toolkit/commands/init/README.md:61]:
```suggestion
When you deploy a contract the function `constructor` is called. If you want to deploy more contracts or pass parameter to the constructor function you have to modify `src/deploy.ts`. Even if you want the default behavior it's cool to take a look at it.
```

File: packages/sc-toolkit/commands/init/assembly/contracts/main.ts
Comment (Thykof) [reply packages/sc-toolkit/commands/init/assembly/contracts/main.ts:25]: it's optional, I can remove it if you think it's not useful

File: packages/sc-toolkit/commands/init/assembly/contracts/main.ts
Comment (AurelienFT) [reply packages/sc-toolkit/commands/init/assembly/contracts/main.ts]: It's a ABI call of massa-as-sdk `()` were missing. The naming is general to be used in other cases as it's an ABI

File: packages/sc-toolkit/commands/init/src/deploy.ts
Comment (AurelienFT) [reply packages/sc-toolkit/commands/init/src/deploy.ts:18]: To match the naming : https://www.geeksforgeeks.org/node-js-__filename-object/

File: packages/sc-deployer/src/index.ts
Comment (AurelienFT) [reply packages/sc-deployer/src/index.ts:60]: It's specific to the usage of OP datastore in this SC it's not generic.

File: packages/sc-toolkit/commands/init/assembly/contracts/main.ts
Comment (AurelienFT) [reply packages/sc-toolkit/commands/init/assembly/contracts/main.ts:11]: I don't know how (in a simple and comprehensive way) you can provide the constructor function and all the possibilities it give without actually writing it. We could have a naming in our SC system saying that if a function is called `constructor` then it can be called only by write access callers but Damir was against putting this kind of "conditions" in our SC system.

**Review by gregLibert — APPROVED** [review]
:+1:

**Pull request merged**
